### ISSUE TITLE
Refund preorder products on failed challenges

### ISSUE THREAD

**Issue by Tirodem** [be-BOP-io-SA/be-BOP#289]
TBD
after #288

tl;dr : to really emulate a crowdfunding behavior, we need to manage the case where the target objective is not reached.
In case of mixed cart during a challenge (let's say donation + physical + preorder), we consider that the donation and the physical are taken in account for the challenge goal current score, but are immediate purchase (as they can be delivered immediately after), so, it's acceptable to not manage automatic refund.
But, as the challenge is supposed to allow future products to exist, it's mandatory to manage auto refund for this kind of products.
So :
- when a product is a "RESOURCE" type
- and it's release date is in the future
- and preorders are enable for this product
- and the product is whitelisted in a challenge (explicit whitelist, not global whitelist with no selected product)
- and the challenge reaches its end date without reaching it's target objective
=> the preorder product is automatically refunded
If the product had a perk discount, we deduce from the refund the discount % applied at the time of the order payment.

Check https://github.com/B2Bitcoin/B2BitcoinBootik/issues/284 to know what we have to refund on SATs transactions:
- fiat value of the product, despite the BTC current value (if it changed since the order payment)
- the same exact amount of SATs

How to refund :
- CB : waiting for Bity specifications
- Lightning : send LNUrlWithdraw to concerned customers through nostr / email / given information at the time of the order
  - resource : https://docs.lnpay.co/api/lnurl-withdraw/how-does-lnurl-withdraw-work
  - is it compatible with perk discount ? As we don't refund the exact order amount, only the preodered product minus perk discount🤔 
- Onchain : send notification asking the customer to send an invoice (give order id, refunded product id, deduced discount)
  - might need to be processed manually

**Comment by Tirodem** [comment]
Will require https://github.com/be-BOP-io-SA/be-BOP/issues/1787
